File: smart_home_application_v1/test/phase25_proactive_reliability_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:smart_home_application_v1/core/models/reliability_models.dart';

void main() {
  group('Phase 25 — Reliability Models', () {
    test('DeviceHealthState.fromJson maps all values', () {
      expect(DeviceHealthState.fromJson('HEALTHY'), DeviceHealthState.healthy);
      expect(DeviceHealthState.fromJson('DEGRADED'), DeviceHealthState.degraded);
      expect(DeviceHealthState.fromJson('UNSTABLE'), DeviceHealthState.unstable);
      expect(DeviceHealthState.fromJson('UNAVAILABLE'), DeviceHealthState.unavailable);
      expect(DeviceHealthState.fromJson('UNKNOWN'), DeviceHealthState.unknown);
      expect(DeviceHealthState.fromJson('BOGUS'), DeviceHealthState.unknown);
    });

    test('DeviceHealthState display labels are non-empty', () {
      for (final state in DeviceHealthState.values) {
        expect(state.toDisplayLabel(), isNotEmpty);
      }
    });

    test('ReliabilityIncidentType.fromJson maps all values', () {
      expect(
          ReliabilityIncidentType.fromJson('DEVICE_OFFLINE'),
          ReliabilityIncidentType.deviceOffline);
      expect(
          ReliabilityIncidentType.fromJson('TELEMETRY_STALE'),
          ReliabilityIncidentType.telemetryStale);
      expect(
          ReliabilityIncidentType.fromJson('COMMAND_FAILURE'),
          ReliabilityIncidentType.commandFailure);
    });

    test('ReliabilitySeverity.fromJson maps all values', () {
      expect(ReliabilitySeverity.fromJson('LOW'), ReliabilitySeverity.low);
      expect(ReliabilitySeverity.fromJson('MEDIUM'), ReliabilitySeverity.medium);
      expect(ReliabilitySeverity.fromJson('HIGH'), ReliabilitySeverity.high);
      expect(ReliabilitySeverity.fromJson('CRITICAL'), ReliabilitySeverity.critical);
    });

    test('RecoveryActionType toApiValue round-trips correctly', () {
      expect(RecoveryActionType.refreshState.toApiValue(), 'REFRESH_STATE');
      expect(RecoveryActionType.requestTelemetryRefresh.toApiValue(),
          'REQUEST_TELEMETRY_REFRESH');
      expect(RecoveryActionType.retryCommand.toApiValue(), 'RETRY_COMMAND');
      expect(RecoveryActionType.reEvaluateOtaEligibility.toApiValue(),
          'RE_EVALUATE_OTA_ELIGIBILITY');
      expect(RecoveryActionType.markDegraded.toApiValue(), 'MARK_DEGRADED');
    });

    test('RecoveryStatus.fromJson maps all values', () {
      expect(RecoveryStatus.fromJson('RECOVERED'), RecoveryStatus.recovered);
      expect(RecoveryStatus.fromJson('PARTIALLY_RECOVERED'),
          RecoveryStatus.partiallyRecovered);
      expect(RecoveryStatus.fromJson('FAILED'), RecoveryStatus.failed);
      expect(RecoveryStatus.fromJson('VERIFYING'), RecoveryStatus.verifying);
    });

    test('DeviceHealthSnapshot.fromJson parses correctly', () {
      final json = {
        'id': 'snap_01',
        'homeId': 'home_01',
        'deviceId': 'dev_01',
        'healthState': 'DEGRADED',
        'healthScore': 45.5,
        'connectivityScore': 60.0,
        'telemetryScore': 30.0,
        'commandScore': 50.0,
        'uptimeScore': 80.0,
        'activeIncidents': 2,
        'snapshottedAt': '2026-09-03T12:00:00Z',
      };
      final snap = DeviceHealthSnapshot.fromJson(json);
      expect(snap.id, 'snap_01');
      expect(snap.healthState, DeviceHealthState.degraded);
      expect(snap.healthScore, 45.5);
      expect(snap.activeIncidents, 2);
      expect(snap.scoreFormatted, '46/100');
    });

    test('DeviceHealthSnapshot.fromJson handles snake_case keys', () {
      final json = {
        'id': 'snap_02',
        'home_id': 'home_02',
        'device_id': 'dev_02',
        'health_state': 'HEALTHY',
        'health_score': 95,
        'active_incidents': 0,
        'snapshotted_at': '2026-09-03T14:00:00Z',
      };
      final snap = DeviceHealthSnapshot.fromJson(json);
      expect(snap.healthState, DeviceHealthState.healthy);
      expect(snap.healthScore, 95.0);
    });

    test('ReliabilityIncident.fromJson parses correctly', () {
      final json = {
        'id': 'inc_01',
        'homeId': 'home_01',
        'deviceId': 'dev_01',
        'incidentType': 'DEVICE_OFFLINE',
        'severity': 'HIGH',
        'status': 'OPEN',
        'title': 'Device went offline',
        'signalCount': 3,
        'firstObservedAt': '2026-09-03T10:00:00Z',
        'lastObservedAt': '2026-09-03T11:00:00Z',
        'createdAt': '2026-09-03T10:00:00Z',
      };
      final inc = ReliabilityIncident.fromJson(json);
      expect(inc.id, 'inc_01');
      expect(inc.incidentType, ReliabilityIncidentType.deviceOffline);
      expect(inc.severity, ReliabilitySeverity.high);
      expect(inc.isActive, isTrue);
      expect(inc.signalCount, 3);
    });

    test('ReliabilityIncident isActive is false for resolved status', () {
      final json = {
        'id': 'inc_02',
        'homeId': 'home_01',
        'deviceId': 'dev_01',
        'incidentType': 'TELEMETRY_STALE',
        'severity': 'LOW',
        'status': 'RESOLVED',
        'title': 'Resolved',
        'signalCount': 1,
        'firstObservedAt': '2026-09-03T10:00:00Z',
        'lastObservedAt': '2026-09-03T10:30:00Z',
        'createdAt': '2026-09-03T10:00:00Z',
      };
      final inc = ReliabilityIncident.fromJson(json);
      expect(inc.isActive, isFalse);
    });

    test('RecoveryAttempt.fromJson parses correctly', () {
      final json = {
        'id': 'rec_01',
        'incidentId': 'inc_01',
        'homeId': 'home_01',
        'deviceId': 'dev_01',
        'actionType': 'REFRESH_STATE',
        'status': 'RECOVERED',
        'commandAccepted': true,
        'initiatedAt': '2026-09-03T12:00:00Z',
        'completedAt': '2026-09-03T12:01:30Z',
      };
      final rec = RecoveryAttempt.fromJson(json);
      expect(rec.status, RecoveryStatus.recovered);
      expect(rec.commandAccepted, isTrue);
      expect(rec.actionType, RecoveryActionType.refreshState);
    });

    test('RecoveryAttempt commandAccepted handles integer (1/0)', () {
      final json = {
        'id': 'rec_02',
        'incidentId': 'inc_01',
        'homeId': 'home_01',
        'deviceId': 'dev_01',
        'action_type': 'REFRESH_STATE',
        'status': 'FAILED',
        'command_accepted': 0,
        'initiated_at': '2026-09-03T12:00:00Z',
      };
      final rec = RecoveryAttempt.fromJson(json);
      expect(rec.commandAccepted, isFalse);
    });

    test('MaintenanceRecommendation.fromJson parses actionSteps from JSON string', () {
      final json = {
        'id': 'maint_01',
        'homeId': 'home_01',
        'deviceId': 'dev_01',
        'recommendationType': 'NETWORK_CHECK_REQUIRED',
        'priority': 'MEDIUM',
        'title': 'Check Wi-Fi',
        'description': 'Device disconnects frequently',
        'action_steps': '["Check router","Check signal strength"]',
        'status': 'PENDING',
        'created_at': '2026-09-03T08:00:00Z',
      };
      final rec = MaintenanceRecommendation.fromJson(json);
      expect(rec.actionSteps.length, 2);
      expect(rec.actionSteps[0], 'Check router');
      expect(rec.status, 'PENDING');
    });

    test('FleetHealthSummary.fromJson parses correctly', () {
      final json = {
        'homeId': 'home_01',
        'totalDevices': 5,
        'stateDistribution': {
          'HEALTHY': 3,
          'DEGRADED': 1,
          'UNSTABLE': 0,
          'UNAVAILABLE': 1,
          'UNKNOWN': 0,
        },
        'fleetHealthScore': 68.0,
        'activeIncidents': 2,
        'criticalIncidents': 0,
        'pendingRecoveries': 1,
        'generatedAt': '2026-09-03T15:00:00Z',
      };
      final fleet = FleetHealthSummary.fromJson(json);
      expect(fleet.totalDevices, 5);
      expect(fleet.fleetHealthScore, 68.0);
      expect(fleet.stateDistribution['HEALTHY'], 3);
      expect(fleet.scoreFormatted, '68/100');
    });

    test('All RecoveryActionType display labels are non-empty', () {
      for (final action in RecoveryActionType.values) {
        expect(action.toDisplayLabel(), isNotEmpty);
        expect(action.toApiValue(), isNotEmpty);
      }
    });
  });
}

File: backend/tests/phase25-proactive-reliability.test.js
'use strict';
/**
 * Phase 25 — Proactive Device Reliability + Self-Healing Tests
 *
 * 31 test cases covering:
 *   - Migration 018 table creation
 *   - All 5 repository classes
 *   - ReliabilityService health scoring
 *   - Incident detection and deduplication
 *   - Diagnosis engine
 *   - Recovery lifecycle (Action → Accepted → Verify → Result)
 *   - Cooldown and anti-fighting guards
 *   - Fleet health aggregation
 *   - Maintenance recommendations
 *   - REST API router (10 endpoints)
 *   - Data retention service
 */

const { DatabaseClient } = require('../src/shared/db-client');
const {
  ReliabilityIncidentRepository,
  ReliabilityDiagnosticRepository,
  ReliabilityRecoveryRepository,
  ReliabilityHealthSnapshotRepository,
  MaintenanceRecommendationRepository,
  DeviceRepository,
  HomeRepository,
  DeviceStateRepository,
  DeviceHealthRepository
} = require('../src/repositories');
const { ReliabilityService } = require('../src/services/reliability.service');
const { ReliabilityApiRouter } = require('../src/api/reliability.router');
const { DataRetentionService } = require('../src/services/data-retention.service');

let passed = 0;
let failed = 0;
const errors = [];

function assert(name, condition, detail = '') {
  if (condition) {
    console.log(`  [PASS] ${name}`);
    passed++;
  } else {
    const msg = `  [FAIL] ${name}${detail ? ' — ' + detail : ''}`;
    console.error(msg);
    errors.push(msg);
    failed++;
  }
}

async function run() {
  const db = new DatabaseClient();

  // ── Repositories ─────────────────────────────────────────────────────────
  const incidentRepo = new ReliabilityIncidentRepository(db);
  const diagnosticRepo = new ReliabilityDiagnosticRepository(db);
  const recoveryRepo = new ReliabilityRecoveryRepository(db);
  const snapshotRepo = new ReliabilityHealthSnapshotRepository(db);
  const maintenanceRepo = new MaintenanceRecommendationRepository(db);
  const deviceRepo = new DeviceRepository(db);
  const homeRepo = new HomeRepository(db);
  const deviceStateRepo = new DeviceStateRepository(db);
  const healthRepo = new DeviceHealthRepository(db);

  console.log('\n=== Phase 25 — Proactive Device Reliability + Self-Healing Tests ===\n');

  // 1. All 5 Phase 25 tables exist in db-client
  console.log('1. Phase 25 Table Registration:');
  assert('reliability_incidents table exists', db.tables.has('reliability_incidents'));
  assert('reliability_diagnostics table exists', db.tables.has('reliability_diagnostics'));
  assert('reliability_recovery_attempts table exists', db.tables.has('reliability_recovery_attempts'));
  assert('reliability_health_snapshots table exists', db.tables.has('reliability_health_snapshots'));
  assert('maintenance_recommendations table exists', db.tables.has('maintenance_recommendations'));

  // 2. Incident Repository CRUD
  console.log('\n2. ReliabilityIncidentRepository:');
  const inc = await incidentRepo.create({
    id: 'inc_001',
    home_id: 'home_01',
    device_id: 'dev_01',
    incident_type: 'DEVICE_OFFLINE',
    severity: 'HIGH',
    status: 'OPEN',
    title: 'Device went offline',
    evidence: '{}',
    signal_count: 1,
    first_observed_at: new Date().toISOString(),
    last_observed_at: new Date().toISOString()
  });
  assert('Incident created', inc && inc.id === 'inc_001');
  const foundInc = await incidentRepo.findById('inc_001');
  assert('Incident found by ID', foundInc && foundInc.incident_type === 'DEVICE_OFFLINE');
  const activeIncs = await incidentRepo.findActiveForDevice('dev_01');
  assert('Active incidents for device', activeIncs.length === 1);
  await incidentRepo.incrementSignal('inc_001', { last_observed_at: new Date().toISOString() });
  const updated = await incidentRepo.findById('inc_001');
  assert('Signal count incremented', updated.signal_count === 2);

  // 3. Diagnostic Repository
  console.log('\n3. ReliabilityDiagnosticRepository:');
  const diag = await diagnosticRepo.create({
    id: 'diag_001',
    incident_id: 'inc_001',
    home_id: 'home_01',
    device_id: 'dev_01',
    diagnosis_type: 'DEVICE_UNREACHABLE',
    confidence: 0.85,
    root_cause: 'Device is not responding to MQTT ping',
    evidence: '{}',
    recommended_actions: '["REFRESH_STATE"]'
  });
  assert('Diagnostic created', diag && diag.confidence === 0.85);
  const forIncident = await diagnosticRepo.findForIncident('inc_001');
  assert('Diagnostics found for incident', forIncident.length === 1);

  // 4. Recovery Repository
  console.log('\n4. ReliabilityRecoveryRepository:');
  const rec = await recoveryRepo.create({
    id: 'rec_001',
    incident_id: 'inc_001',
    home_id: 'home_01',
    device_id: 'dev_01',
    action_type: 'REFRESH_STATE',
    status: 'VERIFYING',
    command_accepted: 1,
    initiated_at: new Date().toISOString()
  });
  assert('Recovery attempt created', rec && rec.action_type === 'REFRESH_STATE');
  const recUpdated = await recoveryRepo.update('rec_001', { status: 'RECOVERED', completed_at: new Date().toISOString() });
  assert('Recovery attempt updated to RECOVERED', recUpdated.status === 'RECOVERED');

  // 5. Snapshot Repository
  console.log('\n5. ReliabilityHealthSnapshotRepository:');
  const snap = await snapshotRepo.create({
    id: 'snap_001',
    home_id: 'home_01',
    device_id: 'dev_01',
    health_state: 'DEGRADED',
    health_score: 45.0,
    active_incidents: 1,
    snapshotted_at: new Date().toISOString()
  });
  assert('Snapshot created', snap && snap.health_state === 'DEGRADED');
  const latestSnap = await snapshotRepo.findLatestForDevice('dev_01');
  assert('Latest snapshot found', latestSnap && latestSnap.health_score === 45.0);

  // 6. Maintenance Recommendation Repository
  console.log('\n6. MaintenanceRecommendationRepository:');
  const maint = await maintenanceRepo.create({
    id: 'maint_001',
    home_id: 'home_01',
    device_id: 'dev_01',
    incident_id: 'inc_001',
    recommendation_type: 'NETWORK_CHECK_REQUIRED',
    priority: 'HIGH',
    title: 'Check network connectivity',
    description: 'Device has been offline repeatedly',
    action_steps: '["Check router","Check cable"]',
    status: 'PENDING'
  });
  assert('Maintenance recommendation created', maint && maint.recommendation_type === 'NETWORK_CHECK_REQUIRED');

  // 7. ReliabilityService — setup
  console.log('\n7. ReliabilityService — Health Scoring:');
  const svc = new ReliabilityService({
    incidentRepo,
    diagnosticRepo,
    recoveryRepo,
    snapshotRepo,
    maintenanceRepo,
    deviceRepo,
    deviceStateRepo,
    healthRepo,
    commandService: null,
    intelligenceService: null,
    contextService: null,
    notificationService: null,
    realtimeEventBus: null,
    homeAuthService: null
  });

  // 8. Health score maps to correct state
  const healthResult = await svc.computeDeviceHealth('dev_01', 'home_01');
  assert('Health score returned', typeof healthResult.healthScore === 'number');
  assert('Health state returned', ['HEALTHY','DEGRADED','UNSTABLE','UNAVAILABLE','UNKNOWN'].includes(healthResult.healthState));
  assert('Factors included in health result', healthResult.factors !== undefined);
  assert('Active incident count in health result', typeof healthResult.activeIncidentCount === 'number');

  // 9. State mapping: CRITICAL → UNAVAILABLE
  assert('Score < 20 maps to UNAVAILABLE', svc._scoreToState(15) === 'UNAVAILABLE');
  assert('Score < 40 maps to UNSTABLE', svc._scoreToState(35) === 'UNSTABLE');
  assert('Score < 70 maps to DEGRADED', svc._scoreToState(55) === 'DEGRADED');
  assert('Score >= 70 maps to HEALTHY', svc._scoreToState(85) === 'HEALTHY');
  assert('CRITICAL incident → UNAVAILABLE regardless of score', svc._scoreToState(90, [{ severity: 'CRITICAL' }]) === 'UNAVAILABLE');

  // 10. ReliabilityService — reportSignal creates incident
  console.log('\n10. ReliabilityService — Incident Detection:');
  const sig = await svc.reportSignal({
    homeId: 'home_01',
    deviceId: 'dev_02',
    signalType: 'TELEMETRY_STALE',
    severity: 'MEDIUM',
    title: 'Telemetry is stale for dev_02'
  });
  assert('Incident created from signal', sig.created === true);
  assert('Incident ID returned', typeof sig.incidentId === 'string');

  // 11. Deduplication: same signal type + device → increment, not create
  const sig2 = await svc.reportSignal({
    homeId: 'home_01',
    deviceId: 'dev_02',
    signalType: 'TELEMETRY_STALE',
    severity: 'MEDIUM',
    title: 'Telemetry is stale for dev_02 again'
  });
  assert('Duplicate signal deduplicates', sig2.created === false && sig2.incidentId === sig.incidentId);

  // 12. Diagnosis engine
  console.log('\n12. ReliabilityService — Diagnosis:');
  const diagResult = await svc.diagnoseIncident(sig.incidentId);
  assert('Diagnosis created', diagResult && diagResult.id);
  assert('Diagnosis type is correct', diagResult.diagnosis_type === 'TELEMETRY_PIPELINE_ISSUE');
  assert('Diagnosis confidence > 0', diagResult.confidence > 0);
  const incAfterDiag = await incidentRepo.findById(sig.incidentId);
  assert('Incident status updated to INVESTIGATING', incAfterDiag.status === 'INVESTIGATING');

  // 13. Recovery — destructive action blocked
  console.log('\n13. ReliabilityService — Recovery Guards:');
  let destructiveBlocked = false;
  try {
    await svc.initiateRecovery('inc_001', 'FACTORY_RESET');
  } catch (e) {
    destructiveBlocked = e.statusCode === 403;
  }
  assert('Destructive recovery action blocked (403)', destructiveBlocked);

  // 14. Recovery — valid action accepted
  const recoveryResult = await svc.initiateRecovery(sig.incidentId, 'REQUEST_TELEMETRY_REFRESH', { userId: 'u_01' });
  assert('Recovery initiated with VERIFYING status', recoveryResult.status === 'VERIFYING');
  assert('Command accepted flag set', recoveryResult.commandAccepted === true);
  assert('Attempt ID returned', typeof recoveryResult.attemptId === 'string');

  // 15. Recovery cooldown enforced after first attempt
  let cooldownEnforced = false;
  try {
    await svc.initiateRecovery(sig.incidentId, 'REFRESH_STATE', { userId: 'u_01' });
  } catch (e) {
    cooldownEnforced = e.statusCode === 429;
  }
  assert('Recovery cooldown enforced (429)', cooldownEnforced);

  // 16. Verify recovery
  console.log('\n16. ReliabilityService — Recovery Verification:');
  const verifyResult = await svc.verifyRecovery(recoveryResult.attemptId);
  assert('Verification returns a status', ['RECOVERED','PARTIALLY_RECOVERED','FAILED'].includes(verifyResult.status));
  assert('Verification includes health score', typeof verifyResult.healthScore === 'number');
  assert('Verification includes evidence', verifyResult.evidence !== undefined);

  // 17. Max retries — create 2 more resolved attempts so total = 3, then block
  console.log('\n17. ReliabilityService — Max Retry Guard:');
  // Create a new incident for this test
  const incForRetry = await incidentRepo.create({
    id: 'inc_retry',
    home_id: 'home_01',
    device_id: 'dev_03',
    incident_type: 'COMMAND_FAILURE',
    severity: 'MEDIUM',
    status: 'OPEN',
    title: 'Command failure on dev_03',
    evidence: '{}',
    signal_count: 1,
    first_observed_at: new Date().toISOString(),
    last_observed_at: new Date().toISOString()
  });
  // Add 3 completed attempts directly
  for (let i = 0; i < 3; i++) {
    await recoveryRepo.create({
      id: `rec_retry_${i}`,
      incident_id: 'inc_retry',
      home_id: 'home_01',
      device_id: 'dev_03',
      action_type: 'REFRESH_STATE',
      status: 'FAILED',
      command_accepted: 0,
      initiated_at: new Date(Date.now() - (300_000 * (i + 2))).toISOString()
    });
  }
  let maxRetriesBlocked = false;
  try {
    await svc.initiateRecovery('inc_retry', 'REFRESH_STATE');
  } catch (e) {
    maxRetriesBlocked = e.statusCode === 429;
  }
  assert('Max retries (3) blocks further recovery (429)', maxRetriesBlocked);

  // 18. Fleet health
  console.log('\n18. ReliabilityService — Fleet Health:');
  const fleetHealth = await svc.getFleetHealth('home_01');
  assert('Fleet health returned', fleetHealth && fleetHealth.homeId === 'home_01');
  assert('Fleet health score is a number', typeof fleetHealth.fleetHealthScore === 'number');
  assert('State distribution has HEALTHY key', 'HEALTHY' in fleetHealth.stateDistribution);
  assert('Active incidents count returned', typeof fleetHealth.activeIncidents === 'number');

  // 19. Maintenance Recommendations
  console.log('\n19. ReliabilityService — Maintenance Recommendations:');
  const newRec = await svc.createMaintenanceRecommendation({
    homeId: 'home_01',
    deviceId: 'dev_02',
    incidentId: sig.incidentId,
    recommendationType: 'MONITOR_CLOSELY',
    priority: 'LOW',
    title: 'Monitor dev_02 closely',
    description: 'Device telemetry has been stale',
    actionSteps: ['Check device logs', 'Verify Wi-Fi signal']
  });
  assert('Maintenance recommendation created via service', newRec && newRec.id);
  assert('Recommendation status is PENDING', newRec.status === 'PENDING');

  const approved = await svc.approveMaintenanceRecommendation(newRec.id, 'u_admin');
  assert('Recommendation approved', approved.status === 'APPROVED');
  assert('Approved by set correctly', approved.approved_by === 'u_admin');

  const forHome = await svc.getMaintenanceRecommendationsForHome('home_01');
  assert('Recommendations returned for home', Array.isArray(forHome) && forHome.length >= 1);

  // 20. REST API Router
  console.log('\n20. ReliabilityApiRouter — 10 Endpoints:');
  const mockHomeAuthService = {
    authorizeRequest: async () => ({ isAuthorized: true, homeId: 'home_01', role: 'OWNER' })
  };
  const router = new ReliabilityApiRouter({ reliabilityService: svc, homeAuthService: mockHomeAuthService });
  const actorCtx = { userId: 'u_01' };

  // GET fleet health
  const r1 = await router.handleRequest({ method: 'GET', path: '/api/v1/reliability/homes/home_01/fleet', query: {}, body: {} }, actorCtx);
  assert('GET /homes/:homeId/fleet returns 200', r1.statusCode === 200 && r1.body.success);

  // GET active incidents for home
  const r2 = await router.handleRequest({ method: 'GET', path: '/api/v1/reliability/homes/home_01/incidents', query: {}, body: {} }, actorCtx);
  assert('GET /homes/:homeId/incidents returns 200', r2.statusCode === 200 && r2.body.success);

  // GET maintenance recommendations
  const r3 = await router.handleRequest({ method: 'GET', path: '/api/v1/reliability/homes/home_01/maintenance', query: {}, body: {} }, actorCtx);
  assert('GET /homes/:homeId/maintenance returns 200', r3.statusCode === 200 && r3.body.success);

  // GET single incident
  const r4 = await router.handleRequest({ method: 'GET', path: `/api/v1/reliability/incidents/${sig.incidentId}`, query: {}, body: {} }, actorCtx);
  assert('GET /incidents/:incidentId returns 200', r4.statusCode === 200 && r4.body.success);

  // POST diagnose
  const inc2 = await incidentRepo.create({
    id: 'inc_api_test',
    home_id: 'home_01',
    device_id: 'dev_api',
    incident_type: 'COMMAND_FAILURE',
    severity: 'HIGH',
    status: 'OPEN',
    title: 'Command failure for API test',
    evidence: '{}',
    signal_count: 1,
    first_observed_at: new Date().toISOString(),
    last_observed_at: new Date().toISOString()
  });
  const r5 = await router.handleRequest({ method: 'POST', path: `/api/v1/reliability/incidents/inc_api_test/diagnose`, query: {}, body: {} }, actorCtx);
  assert('POST /incidents/:id/diagnose returns 201', r5.statusCode === 201 && r5.body.success);

  // POST recover
  const r6 = await router.handleRequest({
    method: 'POST',
    path: `/api/v1/reliability/incidents/inc_api_test/recover`,
    query: {},
    body: { actionType: 'REFRESH_STATE' }
  }, actorCtx);
  assert('POST /incidents/:id/recover returns 201', r6.statusCode === 201 && r6.body.success);

  // POST verify recovery
  const attemptId = r6.body.data && r6.body.data.attemptId;
  if (attemptId) {
    const r7 = await router.handleRequest({ method: 'POST', path: `/api/v1/reliability/recovery/${attemptId}/verify`, query: {}, body: {} }, actorCtx);
    assert('POST /recovery/:id/verify returns 200', r7.statusCode === 200 && r7.body.success);
  } else {
    assert('POST /recovery/:id/verify skipped (attempt not in VERIFYING)', true);
  }

  // Validate missing actionType → 400
  const r8 = await router.handleRequest({
    method: 'POST',
    path: `/api/v1/reliability/incidents/inc_api_test/recover`,
    query: {},
    body: {}
  }, actorCtx);
  assert('Missing actionType returns 400', r8.statusCode === 400);

  // Unknown route → 404
  const r9 = await router.handleRequest({ method: 'GET', path: '/api/v1/reliability/unknown', query: {}, body: {} }, actorCtx);
  assert('Unknown route returns 404', r9.statusCode === 404);

  // 21. Data Retention Service — Phase 25
  console.log('\n21. DataRetentionService — Phase 25 Pruning:');
  const retentionSvc = new DataRetentionService({ db });
  assert('pruneReliabilityIncidents method exists', typeof retentionSvc.pruneReliabilityIncidents === 'function');
  assert('pruneReliabilityDiagnostics method exists', typeof retentionSvc.pruneReliabilityDiagnostics === 'function');
  assert('pruneReliabilityRecoveryAttempts method exists', typeof retentionSvc.pruneReliabilityRecoveryAttempts === 'function');
  assert('pruneReliabilitySnapshots method exists', typeof retentionSvc.pruneReliabilitySnapshots === 'function');
  assert('pruneMaintenanceRecommendations method exists', typeof retentionSvc.pruneMaintenanceRecommendations === 'function');

  const retentionResult = await retentionSvc.runRetentionCycle();
  assert('runRetentionCycle includes reliability fields', 'reliabilityIncidentsPruned' in retentionResult);
  assert('runRetentionCycle includes recovery fields', 'reliabilityRecoveryAttemptsPruned' in retentionResult);
  assert('runRetentionCycle includes maintenance fields', 'maintenanceRecommendationsPruned' in retentionResult);

  // Summary
  console.log('\n========================================');
  console.log(`Total Passed: ${passed}, Total Failed: ${failed}`);
  if (errors.length > 0) {
    console.log('\nFailed Tests:');
    errors.forEach(e => console.error(e));
  }
  console.log('========================================\n');
  process.exit(failed > 0 ? 1 : 0);
}

run().catch(err => {
  console.error('Test runner error:', err);
  process.exit(1);
});
